File: .sentinel/scripts/orch/agents.mjs
#!/usr/bin/env node
// @ts-check
/* ProducedBy=BUILDER RulesHash=BUILDER@1.0 Decision=D-0011 */
/**
 * Agent registry helper used by the prompt renderer.
 * - Enumerates `.sentinel/agents/**`
 * - Validates agent.json + ROLE/PLAYBOOK markdown files
 * - Surfaces normalized metadata (role/playbook contents, mount paths, keywords)
 */

import fs from "node:fs/promises";
import path from "node:path";
import { fileURLToPath } from "node:url";

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);
const ROOT = path.resolve(__dirname, "../../..");
const AGENTS_DIRNAME = ".sentinel/agents";

const REQUIRED_STRING_FIELDS = ["id", "name", "rulesHash", "summary"];
const REQUIRED_ARRAY_FIELDS = ["mount_paths", "prompt_files"];

/**
 * @typedef {Object} AgentPrompt
 * @property {string} path Relative path to the prompt file
 * @property {string} content Markdown content
 */

/**
 * @typedef {Object} AgentMetadata
 * @property {string} id
 * @property {string} name
 * @property {string} rulesHash
 * @property {string} summary
 * @property {string[]} routingKeywords
 * @property {string[]} mountPaths
 * @property {AgentPrompt[]} prompts
 * @property {AgentPrompt} role
 * @property {AgentPrompt} playbook
 * @property {string} dir Relative path to the agent directory
 */

/**
 * Load all agents under `.sentinel/agents/**`.
 * @param {{root?: string}} [options]
 * @returns {Promise<{ agents: AgentMetadata[]; lookup: Map<string, AgentMetadata> }>}
 */
export async function loadAgents(options = {}) {
  const root = options.root ? path.resolve(options.root) : ROOT;
  const agentsDir = path.resolve(root, AGENTS_DIRNAME);
  let dirents;
  try {
    dirents = await fs.readdir(agentsDir, { withFileTypes: true });
  } catch (error) {
    throw new Error(
      `Unable to read agent directory '${relativePath(root, agentsDir)}': ${formatError(error)}`
    );
  }

  const agents = [];
  const errors = [];
  for (const entry of dirents) {
    if (!entry.isDirectory()) continue;
    try {
      const agent = await loadAgentDirectory({ root, agentsDir, agentName: entry.name });
      agents.push(agent);
    } catch (error) {
      errors.push(`[${entry.name}] ${formatError(error)}`);
    }
  }

  if (errors.length) {
    throw new Error(`Agent discovery failed:\n${errors.join("\n")}`);
  }

  agents.sort((a, b) => a.id.localeCompare(b.id, undefined, { sensitivity: "base" }));
  const lookup = new Map(agents.map((agent) => [agent.id.toLowerCase(), agent]));
  return { agents, lookup };
}

/**
 * @param {{root: string; agentsDir: string; agentName: string}} params
 * @returns {Promise<AgentMetadata>}
 */
async function loadAgentDirectory({ root, agentsDir, agentName }) {
  const agentDir = path.join(agentsDir, agentName);
  const relativeDir = relativePath(root, agentDir);
  const agentJsonPath = path.join(agentDir, "agent.json");
  const agentConfig = await readJson(agentJsonPath);

  const missingFields = REQUIRED_STRING_FIELDS.filter((field) => !isNonEmptyString(agentConfig[field]));
  if (missingFields.length) {
    throw new Error(`agent.json missing required string fields: ${missingFields.join(", ")}`);
  }

  const arrayIssues = REQUIRED_ARRAY_FIELDS.filter(
    (field) => !Array.isArray(agentConfig[field]) || !agentConfig[field].every(isNonEmptyString)
  );
  if (arrayIssues.length) {
    throw new Error(`agent.json fields must be string arrays: ${arrayIssues.join(", ")}`);
  }

  if (agentConfig.id !== agentName) {
    throw new Error(`agent id '${agentConfig.id}' must match directory name '${agentName}'`);
  }

  const promptFiles = /** @type {string[]} */ (agentConfig.prompt_files);
  const prompts = [];
  for (const relPath of promptFiles) {
    const absPath = path.resolve(root, relPath);
    const content = await readUtf8(absPath);
    prompts.push({
      path: normalizePath(relativePath(root, absPath)),
      content: content.trim()
    });
  }

  const rolePrompt = findPrompt(prompts, "ROLE");
  const playbookPrompt = findPrompt(prompts, "PLAYBOOK");
  if (!rolePrompt || !playbookPrompt) {
    throw new Error("prompt_files must include ROLE.md and PLAYBOOK.md entries");
  }

  return {
    id: agentConfig.id,
    name: agentConfig.name,
    rulesHash: agentConfig.rulesHash,
    summary: agentConfig.summary,
    routingKeywords: Array.isArray(agentConfig.routing_keywords)
      ? agentConfig.routing_keywords.filter(isNonEmptyString)
      : [],
    mountPaths: /** @type {string[]} */ (agentConfig.mount_paths).map((p) => normalizePath(p)),
    prompts,
    role: rolePrompt,
    playbook: playbookPrompt,
    dir: normalizePath(relativeDir)
  };
}

/**
 * @param {string} p
 */
async function readJson(p) {
  try {
    const raw = await fs.readFile(p, "utf8");
    return JSON.parse(raw);
  } catch (error) {
    throw new Error(`Failed to read/parse JSON '${p}': ${formatError(error)}`);
  }
}

/**
 * @param {string} p
 */
async function readUtf8(p) {
  try {
    return await fs.readFile(p, "utf8");
  } catch (error) {
    throw new Error(`Unable to read '${p}': ${formatError(error)}`);
  }
}

/**
 * @param {unknown} value
 */
function isNonEmptyString(value) {
  return typeof value === "string" && value.trim().length > 0;
}

/**
 * @param {string} root
 * @param {string} target
 */
function relativePath(root, target) {
  return path.relative(root, target);
}

/**
 * @param {string} p
 */
function normalizePath(p) {
  return p.split(path.sep).join("/");
}

/**
 * @param {AgentPrompt[]} prompts
 * @param {string} needle
 * @returns {AgentPrompt | undefined}
 */
function findPrompt(prompts, needle) {
  return prompts.find((prompt) => prompt.path.toLowerCase().includes(needle.toLowerCase()));
}

/**
 * @param {unknown} error
 */
function formatError(error) {
  return error instanceof Error ? error.message : String(error);
}

export default {
  loadAgents
};

File: sentinelkit/sentinelkit/prompt/agents.py
"""Agent registry utilities."""

from __future__ import annotations

from dataclasses import dataclass
import json
from pathlib import Path

from sentinelkit.utils.errors import SentinelKitError, build_error_payload

__all__ = ["AgentRegistry", "load_agents"]


class AgentRegistryError(SentinelKitError):
    """Raised when agent loading fails."""


@dataclass(slots=True)
class AgentMetadata:
    id: str
    name: str
    rules_hash: str
    summary: str
    routing_keywords: list[str]
    mount_paths: list[str]
    role: str
    playbook: str


@dataclass(slots=True)
class AgentRegistry:
    agents: list[AgentMetadata]
    lookup: dict[str, AgentMetadata]


def load_agents(*, root: Path | str) -> AgentRegistry:
    root_path = Path(root).resolve()
    agents_dir = root_path / ".sentinel/agents"
    if not agents_dir.exists():
        raise AgentRegistryError(
            build_error_payload(
                code="agents.missing_dir",
                message=f"Agent directory '{agents_dir}' does not exist.",
            )
        )

    agents: list[AgentMetadata] = []
    errors: list[str] = []
    for entry in sorted(agents_dir.iterdir()):
        if not entry.is_dir():
            continue
        try:
            agents.append(_load_agent(entry, root_path))
        except AgentRegistryError as error:
            errors.append(f"[{entry.name}] {error.payload.message}")

    if errors:
        raise AgentRegistryError(
            build_error_payload(code="agents.load_failed", message="; ".join(errors))
        )

    lookup = {agent.id.lower(): agent for agent in agents}
    return AgentRegistry(agents=agents, lookup=lookup)


def _load_agent(agent_dir: Path, root: Path) -> AgentMetadata:
    config_path = agent_dir / "agent.json"
    try:
        config = json.loads(config_path.read_text(encoding="utf-8"))
    except OSError as error:
        raise AgentRegistryError(
            build_error_payload(
                code="agents.read_json",
                message=f"Unable to read '{config_path}': {error}",
            )
        ) from error
    except json.JSONDecodeError as error:
        raise AgentRegistryError(
            build_error_payload(
                code="agents.parse_json",
                message=f"Failed to parse '{config_path}': {error}",
            )
        ) from error

    required_strings = ["id", "name", "rulesHash", "summary"]
    missing = [field for field in required_strings if not _is_non_empty_string(config.get(field))]
    if missing:
        raise AgentRegistryError(
            build_error_payload(
                code="agents.missing_fields",
                message=f"agent.json missing fields: {', '.join(missing)}",
            )
        )

    mount_paths = config.get("mount_paths") or []
    prompt_files = config.get("prompt_files") or []
    if not isinstance(mount_paths, list) or not all(_is_non_empty_string(item) for item in mount_paths):
        raise AgentRegistryError(
            build_error_payload(code="agents.mounts_invalid", message="mount_paths must be an array of strings.")
        )
    if not isinstance(prompt_files, list) or not all(_is_non_empty_string(item) for item in prompt_files):
        raise AgentRegistryError(
            build_error_payload(code="agents.prompts_invalid", message="prompt_files must be an array of strings.")
        )

    prompts = [_read_prompt(root / path) for path in prompt_files]
    role = _find_prompt(prompts, "ROLE.md")
    playbook = _find_prompt(prompts, "PLAYBOOK.md")
    if role is None or playbook is None:
        raise AgentRegistryError(
            build_error_payload(
                code="agents.prompt_missing",
                message="ROLE.md and PLAYBOOK.md must be listed in prompt_files.",
            )
        )

    routing_keywords = config.get("routing_keywords") or []
    if not isinstance(routing_keywords, list):
        routing_keywords = []

    return AgentMetadata(
        id=config["id"],
        name=config["name"],
        rules_hash=config["rulesHash"],
        summary=config["summary"],
        routing_keywords=[kw for kw in routing_keywords if _is_non_empty_string(kw)],
        mount_paths=[_normalize_path(path) for path in mount_paths],
        role=role,
        playbook=playbook,
    )


def _read_prompt(path: Path) -> tuple[str, str]:
    try:
        return (path.name, path.read_text(encoding="utf-8").strip())
    except OSError as error:
        raise AgentRegistryError(
            build_error_payload(
                code="agents.prompt_io",
                message=f"Unable to read prompt '{path}': {error}",
            )
        ) from error


def _find_prompt(prompts: list[tuple[str, str]], needle: str) -> str | None:
    needle_lower = needle.lower()
    for name, content in prompts:
        if name.lower().endswith(needle_lower):
            return content
    return None


def _normalize_path(path: str) -> str:
    return Path(path).as_posix()


def _is_non_empty_string(value: object) -> bool:
    return isinstance(value, str) and value.strip() != ""
